src/meta/src/rpc/server.rs

```diff
@@ -132,34 +106,56 @@ pub async fn rpc_serve(
     }
 }
 
+fn node_is_leader(leader_rx: &WatchReceiver<(MetaLeaderInfo, bool)>) -> bool {
+    leader_rx.borrow().clone().1
+}
+
 pub async fn rpc_serve_with_store<S: MetaStore>(
     meta_store: Arc<S>,
     address_info: AddressInfo,
     max_heartbeat_interval: Duration,
     lease_interval_secs: u64,
     opts: MetaOpts,
-) -> MetaResult<(JoinHandle<()>, Sender<()>)> {
-    let (current_leader_info, election_handle, election_shutdown, mut leader_rx) = run_elections(
+) -> MetaResult<(JoinHandle<()>, WatchSender<()>)> {
+    // Initialize managers.
+    let (_, election_handle, election_shutdown, mut leader_rx) = run_elections(
         address_info.listen_addr.clone().to_string(),
         meta_store.clone(),
         lease_interval_secs,
     )
     .await?;
 
-    let prometheus_endpoint = opts.prometheus_endpoint.clone();
+    let mut services_leader_rx = leader_rx.clone();
+    let mut note_status_leader_rx = leader_rx.clone();
 
-    // wait until initial election is done
-    if leader_rx.changed().await.is_err() {
-        panic!("Issue receiving leader value from channel");
-    }
+    // FIXME: add fencing mechanism
+    // https://github.com/risingwavelabs/risingwave/issues/6786
 
     // print current leader/follower status of this node
-    let mut note_status_leader_rx = leader_rx.clone();
     tokio::spawn(async move {
-        let span = tracing::span!(tracing::Level::INFO, "node_status");
-        let _enter = span.enter();
+        let _ = tracing::span!(tracing::Level::INFO, "node_status").enter();
+        let mut was_leader = false;
         loop {
-            let (leader_addr, is_leader) = note_status_leader_rx.borrow().clone();
+            if note_status_leader_rx.changed().await.is_err() {
+                tracing::error!("Leader sender dropped");
+                return;
+            }
+
+            let (leader_info, is_leader) = note_status_leader_rx.borrow().clone();
+            let leader_addr = leader_info_to_host_addr(leader_info);
+
+            // Implementation of naive fencing mechanism:
+            // leader nodes should panic if they loose their leader position
+            if was_leader {
+                // && !is_leader {
+                // TODO: enable this again to give leader chance to claim lease again?
+                // I would say no
```

Comment:
IMHO the expected behaviour should be that fencing kicks in if the leader looses the lease, even if it re-acquires it again later. What do you think @yezizp2012 ?

Reply:
Agreed. Let it panic for now.